srml/balances/src/lib.rs

```diff
@@ -437,29 +646,63 @@ where
 		<ReservedBalance<T, I>>::get(who)
 	}
 
-	fn slash(who: &T::AccountId, value: Self::Balance) -> Option<Self::Balance> {
+	fn withdraw(
+		who: &T::AccountId,
+		value: Self::Balance,
+		reason: WithdrawReason,
+		liveness: ExistenceRequirement,
+	) -> result::Result<Self::NegativeImbalance, &'static str> {
+		if let Some(new_balance) = Self::free_balance(who).checked_sub(&value) {
+			if liveness == ExistenceRequirement::KeepAlive && new_balance < Self::existential_deposit() {
+				return Err("payment would kill account")
+			}
+			Self::ensure_account_can_withdraw(who, value, reason, new_balance)?;
+			Self::set_free_balance(who, new_balance);
+			Ok(NegativeImbalance(value))
+		} else {
+			Err("too few free funds in account")
+		}
+	}
+
+	fn slash(
+		who: &T::AccountId,
+		value: Self::Balance
+	) -> (Self::NegativeImbalance, Self::Balance) {
 		let free_balance = Self::free_balance(who);
 		let free_slash = cmp::min(free_balance, value);
 		Self::set_free_balance(who, free_balance - free_slash);
-		Self::decrease_total_stake_by(free_slash);
 		if free_slash < value {
-			Self::slash_reserved(who, value - free_slash)
+			let remaining_slash = value - free_slash;
+			let reserved_balance = Self::reserved_balance(who);
+			let reserved_slash = cmp::min(reserved_balance, remaining_slash);
+			Self::set_reserved_balance(who, reserved_balance - reserved_slash);
+			if reserved_slash < remaining_slash {
+				(NegativeImbalance(free_slash + reserved_slash), remaining_slash - reserved_slash)
+			} else {
+				(NegativeImbalance(value), Zero::zero())
+			}
 		} else {
-			None
+			(NegativeImbalance(value), Zero::zero())
 		}
 	}
 
-	fn reward(who: &T::AccountId, value: Self::Balance) -> result::Result<(), &'static str> {
+	fn deposit_into_existing(
+		who: &T::AccountId,
+		value: Self::Balance
+	) -> result::Result<Self::PositiveImbalance, &'static str> {
 		if Self::total_balance(who).is_zero() {
 			return Err("beneficiary account must pre-exist");
 		}
 		Self::set_free_balance(who, Self::free_balance(who) + value);
-		Self::increase_total_stake_by(value);
-		Ok(())
+		Ok(PositiveImbalance(value))
 	}
 
-	fn increase_free_balance_creating(who: &T::AccountId, value: Self::Balance) -> UpdateBalanceOutcome {
-		Self::set_free_balance_creating(who, Self::free_balance(who) + value)
+	fn deposit_creating(
+		who: &T::AccountId,
+		value: Self::Balance,
+	) -> Self::PositiveImbalance {
+		Self::set_free_balance_creating(who, Self::free_balance(who) + value);
+		PositiveImbalance(value)
```

Comment:
Isn't there a case where `value` is below the existential deposit and `set_free_balance_creating` does nothing but then `PositiveImbalance` adds `value` to the total issuance, therefor putting them out of sync? I think you need a match on the returned enum and only call `PositiveImbalance` on `UpdateBalanceOutcome::Updated`.

Reply:
free_balance is already guaranteed not to be below existential_deposit

Reply:
In which case it would return zero, so `Self::free_balance(who) + value == value`, and there is no reason `value` should need to be >= ed.